openedx/core/djangoapps/discussions/docs/decisions/0001-discussion-configuration-api.rst

```diff
@@ -0,0 +1,65 @@
+Discussion provider configuration
+=================================
+
+Status
+------
+
+Proposal
+
+
+Context
+-------
+
+During the development of discussions plugins, a need arose for a system of
+configuration that would allow admins to configure connections to different
+discussion tools, and have course authors select an established configuration
+without requiring admins to share any secrets with course authors.
+
+For example, a discussion tool provider might require admins to specify OAuth2
+credentials that will be needed to integrate their tool with the platform. This
+needs to be configured by an admin, and used by a course, but ideally course
+authors should not have access to this information.
+
+It would also be useful to have multiple sets of configuration for different
+discussion tools. A course creator can then select one of the available
+configurations based on their requirements.
+
+For instance an organisation might be using a particular Discourse instance
+across a number of their courses. So an admin will create a new configuration
+object that specifies all the settings the Discourse plugin needs to operate,
+such as the site location, the auth credentials etc. When setting up a new
+course, a course author will then be able to see this as one of the possible
+options when configuring Discourse as the discussion tool.
+
+Additionally, we might want to prevent certain configurations to be available
+to all courses on a site. So it would be useful to limit configuration options
+to a specific site, or Organization.
+
+This ADR proposes a new configuration system that allows admins to create
+pre-populated configurations for specific discussions providers, which can then
+be used and optionally overridden by course authors/admins.
+
+Decision
+--------
+
+We can create two models, `DiscussionProviderConfig` and
+`LearningContextDiscussionConfig` that together provide the functionality
+described above.
+
+The `DiscussionProviderConfig` model stores a configuration name (this will be
+visible to course authors configuring a discussion tool), a provider (this will
+be the id of the plugin for which the configuration is defined), and a config
+JSON field (this will store the actual config). A configuration can be also be
```

Comment:
(This may be/probably should be beyond the scope of this PR... just something I was thinking about as I read)

Reply:
Currently, there will only be the django admin site that can be edited as a raw text field. It will be validated as JSON but not whether it matches the scheme expected by a discussion app. 

That said, once we have a plugin API, the validation can be offloaded to the plugin via a standard API, if the plugin supports it. i.e. the plugin can expose a `validate_config` function that if present will be used to check if the configuration is valid.

There will also be a UI to edit these settings in the course-authoring MFE, however that will only allow editing the configuration specific to a course, not the configuration available globally. That said, once the UI already exists, it could be reused to create an admin UI in course-authoring that allows editing these base configuration sets in the future.

Reply:
This all makes sense to me, thanks!